filtering.py

```diff
@@ -0,0 +1,941 @@
+
+    """
+    nb_channels = R_j.shape[1]
+
+    # computes multichannel Wiener gain as v_j R_j inv_Cxx
+    G = torch.zeros_like(inv_Cxx)
+    temp = torch.zeros_like(G[..., 0, 0, :])
+    for (i1, i2, i3) in itertools.product(*(range(nb_channels),)*3):
+        G[..., i1, i2, :] = (G[..., i1, i2, :]
+                             + _mul(R_j[None, :, i1, i3, :],
+                                    inv_Cxx[..., i3, i2, :], temp))
+    G = G * v_j[..., None, None, None]
+    return G
+
+
+def apply_filter(x, W):
+    """
+    Applies a filter on the mixture. Just corresponds to a matrix
+    multiplication.
+
+    Parameters
+    ----------
+    x: torch.Tensor [shape=(nb_frames, nb_bins, nb_channels, 2)]
+        STFT of the signal on which to apply the filter.
+
+    W: torch.Tensor [shape=(nb_frames, nb_bins, nb_channels, nb_channels, 2)]
+        filtering matrices, as returned, e.g. by :func:`wiener_gain`
+
+    Returns
+    -------
+    y_hat: torch.Tensor [shape=(nb_frames, nb_bins, nb_channels, 2)]
+        filtered signal
+    """
+    nb_channels = W.shape[-2]
+
+    # apply the filter
+    y_hat = torch.zeros_like(x)
+    temp = torch.zeros_like(W[..., 0, :])
+    for i in range(nb_channels):
+        y_hat = y_hat + _mul(W[..., i, :], x[..., i, None, :], temp)
+    return y_hat
+
+
+def get_mix_model(v, R):
+    """
+    Compute the model covariance of a mixture based on local Gaussian models.
+    simply adds up all the v[..., j] * R[..., j]
+
+    Parameters
+    ----------
+    v: torch.Tensor [shape=(nb_frames, nb_bins, nb_sources)]
+        Power spectral densities for the sources
+
+    R: torch.Tensor [shape=(nb_bins, nb_channels, nb_channels, 2, nb_sources)]
+        Spatial covariance matrices of all sources
+
+    Returns
+    -------
+    Cxx: torch.Tensor [shape=(nb_frames, nb_bins, nb_channels, nb_channels, 2)]
+        Covariance matrix for the mixture
+    """
+    nb_channels = R.shape[1]
+    (nb_frames, nb_bins, nb_sources) = v.shape
+    Cxx = torch.zeros((nb_frames, nb_bins, nb_channels, nb_channels, 2),
+                      dtype=R.dtype, device=v.device)
+    for j in range(nb_sources):
+        Cxx = Cxx + v[..., j, None, None, None] * R[None, ..., j]
+    return Cxx
+
+
+def _covariance(y_j):
+    """
+    Compute the empirical covariance for a source.
+
+    Parameters
+    ----------
+    y_j: torch.Tensor [shape=(nb_frames, nb_bins, nb_channels, 2)].
+          complex stft of the source.
+
+    Returns
+    -------
+    Cj: torch.Tensor [shape=(nb_frames, nb_bins, nb_channels, nb_channels, 2)]
+        just y_j * conj(y_j.T): empirical covariance for each TF bin.
+    """
+    (nb_frames, nb_bins, nb_channels) = y_j.shape[:-1]
+    Cj = torch.zeros((nb_frames, nb_bins, nb_channels, nb_channels, 2),
+                     dtype=y_j.dtype,
+                     device=y_j.device)
+    temp = torch.zeros_like(y_j[..., 0, :])
+    for (i1, i2) in itertools.product(*(range(nb_channels),)*2):
```

Comment:
use `torch.cartesian_prod(tensor_a, tensor_b)`

Reply:
no it doesn't handle compex values

Reply:
@aliutkus but so doesn't itertools or maybe I don't get the code ;-)

Reply:
But so doesn't itertools or maybe I don't get this?

Reply:
haaaa, I get your comment now. I thought you were talking of the actual computation being done here, which involves complex values

Reply:
:+1: